modules/common/memcached/memcached.go

```diff
@@ -0,0 +1,44 @@
+package memcached
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	memcachedv1 "github.com/openstack-k8s-operators/infra-operator/apis/memcached/v1beta1"
+	oko_svc "github.com/openstack-k8s-operators/lib-common/modules/common/service"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8s_labels "k8s.io/apimachinery/pkg/labels"
+)
+
+// NewMemcached function will return a memcached v1beta1 instance of the CR. This can then be used
+// with controllerutil.CreateOrPatch() to create the resource in the cluster.
+func NewMemcached(name string, namespace string) *memcachedv1.Memcached {
+	return &memcachedv1.Memcached{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "memcached.openstack.org/v1beta1",
+			Kind: "Memcached",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Namespace: namespace,
+		},
+	}
+}
+
+// GetMemcachedServiceName function searches for all memcached services matching the provided memcachedName
+func GetMemcachedServiceName(ctx context.Context, memcachedName string, h *helper.Helper, namespace string) (svc string, err error) {
+	labelSelector := k8s_labels.Set{"app": "memcached"}
```

Comment:
Currently the memcached controller creates a series of pod for a memcached CR, each with unique labels to distinguis between two CR, e.g.:
```oc get pod/memcached-0 -o json | jq .metadata.labels
{
  "app": "memcached",
  "controller-revision-hash": "memcached-67b5546d47",
  "cr": "memcached-memcached",
  "memcached/name": "memcached",
  "memcached/namespace": "openstack",
  "memcached/uid": "96adb626-96a4-40d1-a1b9-f3047f0137d7",
  "owner": "infra-operator",
  "statefulset.kubernetes.io/pod-name": "memcached-0"
}
```
So the labelSelector should also filter for memcached/name: memcachedName.

Reply:
So, to be clear. This function is just getting the headless service. Which there will only ever be one of regardless of the number of replicas:

```
❯ oc get memcached horizon-memcached -o jsonpath='{.spec.replicas}' | jq
3

❯ oc get po -l app=memcached
NAME                  READY   STATUS              RESTARTS   AGE
horizon-memcached-0   1/1     Running             0          65s
horizon-memcached-1   1/1     Running             0          25s
horizon-memcached-2   0/1     ContainerCreating   0          19s

❯ oc get svc -l app=memcached
NAME                TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)     AGE
horizon-memcached   ClusterIP   None         <none>        11211/TCP   71s

```

So, I think the best course of action would be for me to add a new function that gets the list of pods, where then I will need to return a slice of strings like `[]String{'memcached-0', 'memcached-1', 'memcached-2'}`. But I think I should do that in a seperate function and leave this here in case we have a use for getting just the SVC like this. (Which btw is currently how I'm doing it in Horizon. I just add the single SVC and rely on whatever mechanism the k8s service uses to send traffic to the pods.

Reply:
That is a good point. Do we anticipate using the service at all then?

Maybe I should scrap the function entirely if it isn't as useful as I first thought. I can just work on getting the pod list instead if that's the case.

Reply:
Based on https://github.com/openstack-k8s-operators/openstack-operator/pull/231, we just instantiate a headless service for getting DNS entries for the pods. So yes, we don't plan on accessing memcached via a service directly.

Reply:
I think we can close this one off too since I removed this function. But thanks again for the pointer on the headless service name resolution. Makes it much easier.